### ISSUE TITLE
Socket::PF_INET6 is nil on 9.3 when ipv6 is not available, unlike with MRI and 9.2

### ISSUE THREAD

**Issue by p-mongo** [jruby/jruby#7069]
Socket::PF_INET6 is nil on 9.3 whereas it was an integer (10) on 9.2.

This might require a specific OS/JRE(?) to reproduce, I reproduced this in Docker using ubuntu 20.04 and openjdk JRE 11. The full log is at the very end.

**Environment Information**
  
Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc)
* jruby 9.3.3.0 (2.6.8) 2022-01-19 b26de1f5c5 OpenJDK 64-Bit Server VM 11.0.13+8-Ubuntu-0ubuntu1.20.04 on 11.0.13+8-Ubuntu-0ubuntu1.20.04 +jit [linux-x86_64]
* no command line flags or options
* Operating system and platform (e.g. `uname -a`)
* ubuntu 20.04 in docker
* Linux f1e690783ce8 5.15.0-2-amd64 #1 SMP Debian 5.15.5-2 (2021-12-18) x86_64 x86_64 x86_64 GNU/Linux

Other relevant info you may wish to add:
* Installed or activated gems - none
* Application/framework version (e.g. Rails, Sinatra) - none
* Environment variables - none

**Expected Behavior**

Reproduce code:
```
irb -rsocket
```
Type in irb:
```
Socket::PF_INET6
```

When I use `Socket::PF_INET6` constant, I expect it to have an integer value. This is the case on JRuby 9.2:

```
f1e690783ce8# rbenv install jruby-9.2.20.1
Downloading jruby-dist-9.2.20.1-bin.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/79cdbc475a7041f4b44766c069051d21dd2473ce1638a0b668aa9439fb631963
Installing jruby-9.2.20.1...
Installed jruby-9.2.20.1 to /root/.rbenv/versions/jruby-9.2.20.1

jruby-9.2.20 => jruby-9.2.20.1
f1e690783ce8# rbenv shell jruby-9.2.20    
f1e690783ce8# irb -rsocket                
irb(main):001:0> Socket::PF_INET6
=> 10
irb(main):002:0> 
```

**Actual Behavior**

On JRuby 9.3 the constant is defined but its value is `nil`:

```
f1e690783ce8# rbenv install jruby-9.3.3.0 
Downloading jruby-dist-9.3.3.0-bin.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/3da828cbe287d5468507f1c2c42bef6cf34bc5361bcd6a5d99c207b21b9fdc5c
Installing jruby-9.3.3.0...
Installed jruby-9.3.3.0 to /root/.rbenv/versions/jruby-9.3.3.0

jruby-9.3.3 => jruby-9.3.3.0
f1e690783ce8# rbenv shell jruby-9.3.3    
f1e690783ce8# irb -rsocket               
/root/.rbenv/versions/jruby-9.3.3.0/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::AF_INET6
/root/.rbenv/versions/jruby-9.3.3.0/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::IPPROTO_ICMPV6
/root/.rbenv/versions/jruby-9.3.3.0/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::IPPROTO_IPV6
/root/.rbenv/versions/jruby-9.3.3.0/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::PF_INET6
/root/.rbenv/versions/jruby-9.3.3.0/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::SOL_IPV6
irb(main):001:0> Socket::PF_INET6
=> nil
irb(main):002:0> 
```

9.3.2.0 is also affected.

Complete testing log: https://gist.github.com/p-mongo/25ce8b6d9d91dbc8dc30f16dbea46c98

**Comment by ahorek** [comment]
is ipv6 usable on that docker platform? if there's no ipv6 interface available we do set ipv6 constants to nil. Since we can't inspect ipv6 capabilities at compile-time, we use a java trick that could behave differently than CRuby.

see https://github.com/jruby/jruby/pull/6945

I guess there could be a problem with mocked connections (for tests), that depends on these constants even if no real interface is usable... the question is, how to detect this reliably on Java.

I noticed a second problem with verbose warnings
```
warning: already initialized constant Socket::Constants::AF_INET6
```
this is definitely an unintentional bug and I think that const_defined? is probably also affected, but this one should be easy to fix.

**Comment by p-mongo** [comment]
> is ipv6 usable on that docker platform?

Perhaps not, which would explain why on my machine the constant was defined.

```
root@1d66935478aa:/# ifconfig 
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.0.2  netmask 255.255.0.0  broadcast 172.17.255.255
        ether 02:42:ac:11:00:02  txqueuelen 0  (Ethernet)
        RX packets 5417  bytes 22539415 (22.5 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4144  bytes 279208 (279.2 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

root@1d66935478aa:/# ifconfig  lo0 inet6
root@1d66935478aa:/# ifconfig  lo0 inet6 ::1
Don't know how to set addresses for family 10.
```


However, this does not stop MRI from providing the v6 constants:

```
root@1d66935478aa:/# ruby -v
ruby 2.7.0p0 (2019-12-25 revision 647ee6f091) [x86_64-linux-gnu]
root@1d66935478aa:/# irb -rsocket
irb(main):001:0> Socket::PF_INET
=> 2
irb(main):002:0> Socket::PF_INET6
=> 10
irb(main):003:0> 
```

If ipv6 is not available, I would expect e.g. socket creation to fail, similarly to the message above when I tried to add ::1 :


```
Don't know how to set addresses for family 10.
```

But, this sort of diagnostics requires the address family to actually be defined.


In the Ruby MongoDB driver we have code like this that attempts to instantiate a socket:
```
::Socket.getaddrinfo(host, nil, family, ::Socket::SOCK_STREAM)
```

... where `family` is set to either INET or INET6. On the affected system, this code produces rather puzzling errors.

MRI:

```
root@1d66935478aa:/# ruby -v
ruby 2.7.0p0 (2019-12-25 revision 647ee6f091) [x86_64-linux-gnu]
root@1d66935478aa:/# irb -rsocket
irb(main):001:0> Socket.new Socket::AF_INET, Socket::SOCK_STREAM
=> #<Socket:fd 5>
irb(main):002:0> Socket.new Socket::AF_INET6, Socket::SOCK_STREAM
=> #<Socket:fd 6>
```

JRuby 9.3:

```
irb(main):005:0> Socket.new Socket::AF_INET, Socket::SOCK_STREAM
=> #<Socket:fd 12>
irb(main):006:0> Socket.new Socket::AF_INET6, Socket::SOCK_STREAM
Traceback (most recent call last):
       11: from /root/.rbenv/versions/jruby-9.3.3/bin/irb:23:in `<main>'
       10: from org/jruby/RubyKernel.java:1052:in `load'
        9: from /root/.rbenv/versions/jruby-9.3.3.0/lib/ruby/gems/shared/gems/irb-1.0.0/exe/irb:11:in `<main>'
        8: from org/jruby/RubyKernel.java:1237:in `catch'
        7: from org/jruby/RubyKernel.java:1237:in `catch'
        6: from org/jruby/RubyKernel.java:1507:in `loop'
        5: from org/jruby/RubyKernel.java:1091:in `eval'
        4: from (irb):6:in `evaluate'
        3: from org/jruby/RubyIO.java:868:in `new'
        2: from org/jruby/RubyClass.java:907:in `new'
        1: from org/jruby/ext/socket/RubySocket.java:192:in `initialize'
TypeError (no implicit conversion from nil to integer)
```

The error suggests the application passed nil somewhere, but I followed the socket creation  code at https://ruby-doc.org/stdlib-2.5.3/libdoc/socket/rdoc/Socket.html and there isn't a nil anywhere in the code I actually typed up.

MRI is also able to create a socket even though there is (perhaps?) no working ipv6 stack on the system.

```
irb(main):004:0> s=Socket.new Socket::AF_INET6, Socket::SOCK_STREAM
irb(main):005:0> s.connect Socket.pack_sockaddr_in(80, 'example.com')
Traceback (most recent call last):
        5: from /usr/bin/irb:23:in `<main>'
        4: from /usr/bin/irb:23:in `load'
        3: from /usr/lib/ruby/gems/2.7.0/gems/irb-1.2.1/exe/irb:11:in `<top (required)>'
        2: from (irb):5
        1: from (irb):5:in `connect'
Errno::EINVAL (Invalid argument - connect(2) for 93.184.216.34:80)
```

So the issues with the 9.3 behavior to me are:

1. It differs from MRI behavior and is also a regression from 9.2 which matches MRI behavior
2. It produces confusing errors

**Comment by ahorek** [comment]
> The error suggests the application passed nil somewhere
because we set them to nil here https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/ext/socket/RubySocket.java#L120

this is wrong because typical guards to ipv6 availability will fail
```Socket.const_defined?(:AF_INET6)``` => returns true, but the value is nil. If nothing else, it should be undefined.

CRuby does define these constants even if they're not really usable and tests depend on a failure during opening a socket
https://github.com/ruby/ruby/blob/master/spec/ruby/library/socket/fixtures/classes.rb#L50

the related PR was mainly about fixing tests on github actions where ipv6 fails the same way as you described. The idea of checking ip6 availability makes sense, but since it caused incompatibility with CRuby's behavior, we should find another way how to avoid these failures.

**Comment by ahorek** [comment]
ok, tests seem to pass without additional changes https://github.com/jruby/jruby/pull/7071

@p-mongo could you test if it works for you?

**Comment by headius** [comment]
Fixed by #7071!

@p-mongo I will generate an updated snapshot build that should show up soon at https://oss.sonatype.org/content/repositories/snapshots/org/jruby/jruby-dist/9.3.4.0-SNAPSHOT/. This should be easier for you to test, since your JRuby build environment seems to have some strange issues.

Please let us know right away if this does not resolve your issue!
